add the ability to conceal passwords and otps when copied to pasteboard

georgesnow opened the pull request:
this adds the ability to hide passwords and OTPs from clipboard managers

zkytech reviewed (changes requested):
Lots of code style problem, your ide should have som hint for you. Before submit , execute `npm run lint` to check code style.

zkytech reviewed (changes requested):
The proteedtedCopy function works fine, but there are some problem:
- deprecated api: your code seems not based on the latest version. I have update many deprecated api in last commit. Check the hints from your ide and execute `npm run lint` before submit.
- duplicate code: `protectedCopy` is defined in both `utils/keepassLoader.ts` and `index.tsx`

zkytech commented on extensions/keepassxc/src/index.tsx:
```diff
@@ -55,33 +57,60 @@ export default function Command() {
+                icon={Icon.TextDocument}
                 onAction={() => {
-                  pastePassword(entry).then(() => closeMainWindow());
+                  copyAndPastePassword(entry).then(() => closeMainWindow());
```
```suggestion
                  copyAndPastePassword(entry).then(() => closeMainWindow()).catch(errorHandler);
```
I have forgot to add errorHandler in the end, it's my mistake, Fix it in this commit.

zkytech commented on extensions/keepassxc/src/index.tsx:
```diff
@@ -55,33 +57,60 @@ export default function Command() {
                 icon={Icon.Clipboard}
                 shortcut={{ modifiers: ["cmd"], key: "enter" }}
                 onAction={() => {
                   copyPassword(entry).then(() => closeMainWindow());
```
```suggestion
                  copyPassword(entry).then(() => closeMainWindow()).catch(errorHandler);
```
I have forgot to add errorHandler in the end, it's my mistake, Fix it in this commit.

zkytech commented on extensions/keepassxc/src/index.tsx:
```diff
@@ -55,33 +57,60 @@ export default function Command() {
                 icon={Icon.Clipboard}
                 shortcut={{ modifiers: ["cmd"], key: "b" }}
                 onAction={() => {
                   copyUsername(entry).then(() => closeMainWindow());
```
```suggestion
                  copyUsername(entry).then(() => closeMainWindow()).catch(errorHandler);
```
I have forgot to add errorHandler in the end, it's my mistake, Fix it in this commit.

zkytech commented on extensions/keepassxc/src/index.tsx:
```diff
@@ -55,33 +57,60 @@ export default function Command() {
+                shortcut={{ modifiers: ["control", "cmd"], key: "c" }}
+                onAction={() => {
+                  // copyOTP(entry).then(() => closeMainWindow());
+                  copyOTP(entry).then(() => closeMainWindow());
```
```suggestion
                  copyOTP(entry).then(() => closeMainWindow()).catch(errorHandler);
```

zkytech reviewed (changes requested):
I have forgot to add errorHandler in the end of chain. Fix it in this commit.

georgesnow commented:
@zkytech Thanks for taking a look. I moved this to a draft request. I had started work on this a while ago. I mistakenly thought I had cleaned it up better 😕. I appreciate you taking the time.  Anyway, I have decent number of changes to fix. I will try and address all of these when I get another block of time.

georgesnow commented:
@zkytech  I think I fixed all the outstanding issues.

zkytech commented:
> @zkytech I think I fixed all the outstanding issues.

To solve the remaining issues, I have made a commit to my fork based on your code, please execute:
1. clone my repo :`git clone https://github.com/zkytech/extensions.git`
2. remove keepassxc directory in your local repo :`rm -rf georgesnow/extensions/extensions/keepassxc`
3. copy keepassxc directory from the directory you just cloned: `copy -r zkytech/extensions/extensions/keepassxc georgesnow/extensions/extensions/`

georgesnow commented:
@zkytech  question why did you remove the error handling for TOTP std out? if the TOTP is triggered accidentally on an entry that doesn't have OTP the extension crashes. 

https://github.com/georgesnow/extensions/blob/0d389dd0cc3c6efbd1b9f73ac0d7b03f314a136c/extensions/keepassxc/src/utils/keepassLoader.ts#L84

zkytech commented:
> @zkytech question why did you remove the error handling for TOTP std out? if the TOTP is triggered accidentally on an entry that doesn't have OTP the extension crashes.
> 
> https://github.com/georgesnow/extensions/blob/0d389dd0cc3c6efbd1b9f73ac0d7b03f314a136c/extensions/keepassxc/src/utils/keepassLoader.ts#L84

https://github.com/zkytech/extensions/blob/ca2ddeb9b00f5ae466b4c8284a62780aaa0fbf34/extensions/keepassxc/src/index.tsx#L91

georgesnow commented:
Correct, but the error handler in your fork removed the condition for keepassxc-cli where an entry doesn't have OTP. If the user triggers the keyboard shortcut for OTP on an entry with no OTP. The return value from the cli is something error handler doesn't have condition for. This causes the extension to crash. 

Can I go ahead and add that back in?

zkytech commented:
> Correct, but the error handler in your fork removed the condition for keepassxc-cli where an entry doesn't have OTP. If the user triggers the keyboard shortcut for OTP on an entry with no OTP. The return value from the cli is something error handler doesn't have condition for. This causes the extension to crash.
> 
> Can I go ahead and add that back in?

Please clone and build my repo to run a test.
It works fine. if condition is not required to handle such error.

georgesnow commented:
Correction I shouldn't have said crash. It logs a failure, but doesn't  it seem a bit more graceful to notify the user with a HUD that an entry is missing OTP? I added it because I had triggered on various entries in, which I thought had OTP setup for. It made me think it would be helpful to know that.

zkytech commented:
> Correction I shouldn't have said crash. It logs a failure, but doesn't it seem a bit more graceful to notify the user with a HUD that an entry is missing OTP? I added it because I had triggered on various entries in my testing, which I thought had OTP setup for. It made me think it would be helpful to know that.
Run a test, you will see Raycast display error message in bottom of UI with red color.

georgesnow commented:
<img width="804" alt="image" src="https://user-images.githubusercontent.com/12751441/192297254-7e45fb9d-5ca4-4ae6-964a-671586011514.png">

It seems easy to miss and if the entry name is long like the one in my database (I have longer ones). The message gets cut off, but if you think that's sufficient I will go ahead merge in the changes.

zkytech commented:
It’s ok

georgesnow commented:
done.

pernielsentikaer commented:
Thanks for an outstanding review @zkytech 🔥

pernielsentikaer commented:
@georgesnow do you want to be added as a contributor to this extension, if so, send me your Raycast @handle 🙂

georgesnow commented:
> @georgesnow do you want to be added as a contributor to this extension, if so, send me your Raycast @handle 🙂

sure, sorry for the delay in responding. It should be the same as my handle for GitHub @georgesnow 

thanks!
